handler/core/src/main/java/io/knotx/fragments/handler/action/cb/CircuitBreakerActionFactory.java

```diff
@@ -0,0 +1,172 @@
+import io.knotx.fragments.handler.api.domain.FragmentResult;
+import io.knotx.fragments.handler.exception.DoActionExecuteException;
+import io.knotx.fragments.handler.exception.DoActionNotDefinedException;
+import io.vertx.circuitbreaker.CircuitBreaker;
+import io.vertx.circuitbreaker.impl.CircuitBreakerImpl;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This is a factory class creating action, which provides circuit breaker mechanism. It protects
+ * the `doAction` action against overloading when it does not respond on time. If t
+ */
+@Cacheable
+public class CircuitBreakerActionFactory implements ActionFactory {
+
+  static final String FALLBACK_TRANSITION = "fallback";
+  static final String FACTORY_NAME = "cb";
+
+  @Override
+  public String getName() {
+    return FACTORY_NAME;
+  }
+
+  @Override
+  public Action create(String alias, JsonObject config, Vertx vertx, Action doAction) {
+    if (isNull(doAction)) {
+      throw new DoActionNotDefinedException("Circuit Breaker action requires `doAction` defined");
+    }
+    CircuitBreakerActionFactoryOptions options = new CircuitBreakerActionFactoryOptions(config);
+    CircuitBreaker circuitBreaker = new CircuitBreakerImpl(options.getCircuitBreakerName(), vertx,
+        options.getCircuitBreakerOptions());
+
+    return new CircuitBreakerAction(circuitBreaker, doAction, alias,
+        fromConfig(options.getLogLevel()));
+  }
+
+  public static class CircuitBreakerAction implements Action {
+
+    public static final String INVOCATION_COUNT_LOG_KEY = "invocationCount";
+    public static final String ERROR_LOG_KEY = "error";
+    private final CircuitBreaker circuitBreaker;
+    private final Action doAction;
+    private final ActionLogLevel actionLogLevel;
+    private final String alias;
+
+    CircuitBreakerAction(CircuitBreaker circuitBreaker, Action doAction, String alias,
+        ActionLogLevel actionLogLevel) {
+      this.circuitBreaker = circuitBreaker;
+      this.doAction = doAction;
+      this.alias = alias;
+      this.actionLogLevel = actionLogLevel;
+    }
+
+    @Override
+    public void apply(FragmentContext fragmentContext,
+        Handler<AsyncResult<FragmentResult>> resultHandler) {
+      AtomicInteger counter = new AtomicInteger();
+      ActionLogger actionLogger = ActionLogger.create(alias, actionLogLevel);
+      circuitBreaker.executeWithFallback(
+          promise -> executeCommand(promise, fragmentContext, counter, actionLogger),
+          throwable -> handleFallback(fragmentContext, throwable, counter, actionLogger)
+      ).setHandler(resultHandler);
+    }
+
+    private void executeCommand(Promise<FragmentResult> promise, FragmentContext fragmentContext,
+        AtomicInteger counter, ActionLogger actionLogger) {
+      counter.incrementAndGet();
```

Comment:
```suggestion
      int attempt = counter.incrementAndGet();
```
And pass the `attempt` to the `handleResult` and `handleFail` methods, as they use they read-only the value.

Reply:
It is not true. `AtomicCounter` variable is passed to "recalculate" the number of invocations just before the circuit breaker action responds with a node log.